### PR TITLE
Cache classes editor help (a.k.a. faster editor startup)

### PR DESCRIPTION
**Pull request by RandomShaper**
~(Built on top of #71644, which is included here. Please review newest commit only.)~

The class reference is now cached and regenerated every time the doc tool is run or the commit hash changes. This avoids, on every regular run of the editor, paying the cost of instantiating all the known classes at startup.

Docs generation have two steps: a synchronous one, which just contributes to the perceived startup time, and one which runs on a thread, whose cost is negligible, but I guess can affect some low performance systems for the brief time it lasts.

On my system (AMD Ryzen 9 5900HX with NVMe SSD), a regular startup gives these times:
- Before this PR: 240 ms sync + 114 ms async.
- With this PR: 0 ms sync + 200 ms async.
**UPDATE:** I want to clarify that the zero milliseconds is not due to the SSD. If the cache is already generated, the only things that needs to happen synchronously is checking if the file exists. Parsing and validating it happens asynchronously.

In short, **on my system the Godot editor takes roughly 1/4 of a second less to start. Not a lot, but good in any case**.

### CODE REVIEW

**Comment by RandomShaper** [comment]
Rebased on top of 4.1.

**Comment by RandomShaper** [comment]
Rebased again, after the rebase of the PR dependency.

**Comment by akien-mga** [comment]
Did some before/after testing, with a naive approach timing the editor startup time on a simple/near empty project.

System:
```
$ inxi -C
CPU:
  Info: quad core model: Intel Core i7-8705G bits: 64 type: MT MCP cache:
    L2: 1024 KiB
  Speed (MHz): avg: 1726 min/max: 800/4100 cores: 1: 900 2: 913 3: 900
    4: 3100 5: 900 6: 900 7: 3100 8: 3100
```
Godot build: `scons target=editor`

For each test I removed `~/.cache/godot`, `~/.local/share/godot/app_userdata/<project name>` and `<project path>/.godot`.
And then I timed the startup of the editor when using the `--quit` argument, so it quits as soon as it's initialized:
```
time ./bin/godot.linuxbsd.editor.x86_64 -e --path ~/New\ Game\ Project4.0/ --quit
```

### Before this PR (68b8156fe3075e0e45b070d26fce2a0baae6f293)

Cold run, generates shader cache, etc. (average of 3 runs after 1 warmup run):
8.765s

Subsequent runs (average of 20, after 3 warmup runs):
```
$ hyperfine -i -w3 -m20 "./bin/godot.linuxbsd.editor.x86_64.old -e --path ~/New\ Game\ Project4.0/ --quit"
Benchmark 1: ./bin/godot.linuxbsd.editor.x86_64.old -e --path ~/New\ Game\ Project4.0/ --quit
  Time (mean ± σ):      6.623 s ±  0.264 s    [User: 5.017 s, System: 0.469 s]
  Range (min … max):    5.559 s …  6.819 s    20 runs
```

### With this PR (rebased on 68b8156f)

Cold run, generates shader cache, etc. + doc cache (average of 3 runs after 1 warmup run):
8.655s

Subsequent runs (average of 20, after 3 warmup runs):
```
hyperfine -i -w3 -m20 "./bin/godot.linuxbsd.editor.x86_64 -e --path ~/New\ Game\ Project4.0/ --quit"
Benchmark 1: ./bin/godot.linuxbsd.editor.x86_64 -e --path ~/New\ Game\ Project4.0/ --quit
  Time (mean ± σ):      6.282 s ±  0.238 s    [User: 4.806 s, System: 0.485 s]
  Range (min … max):    5.332 s …  6.459 s    20 runs
```

So it does seem to shave off around 350 ms in average for me, though it's not a very reliable test. Interesting that the cold run also seems to be tangentially faster / same duration, so saving the cache doesn't seem to be slow.

**Comment by akien-mga** [comment]
Did some before/after testing, with a naive approach timing the editor startup time on a simple/near empty project.

System:
```
$ inxi -C
CPU:
  Info: quad core model: Intel Core i7-8705G bits: 64 type: MT MCP cache:
    L2: 1024 KiB
  Speed (MHz): avg: 1726 min/max: 800/4100 cores: 1: 900 2: 913 3: 900
    4: 3100 5: 900 6: 900 7: 3100 8: 3100
```
Godot build: `scons target=editor`

For each test I removed `~/.cache/godot`, `~/.local/share/godot/app_userdata/<project name>` and `<project path>/.godot`.
And then I timed the startup of the editor when using the `--quit` argument, so it quits as soon as it's initialized:
```
time ./bin/godot.linuxbsd.editor.x86_64 -e --path ~/New\ Game\ Project4.0/ --quit
```

### Before this PR (68b8156fe3075e0e45b070d26fce2a0baae6f293)

Cold run, generates shader cache, etc. (average of 3 runs after 1 warmup run):
8.765s

Subsequent runs (average of 20, after 3 warmup runs):
```
$ hyperfine -i -w3 -m20 "./bin/godot.linuxbsd.editor.x86_64.old -e --path ~/New\ Game\ Project4.0/ --quit"
Benchmark 1: ./bin/godot.linuxbsd.editor.x86_64.old -e --path ~/New\ Game\ Project4.0/ --quit
  Time (mean ± σ):      6.623 s ±  0.264 s    [User: 5.017 s, System: 0.469 s]
  Range (min … max):    5.559 s …  6.819 s    20 runs
```

### With this PR (rebased on 68b8156f)

Cold run, generates shader cache, etc. + doc cache (average of 3 runs after 1 warmup run):
8.655s

Subsequent runs (average of 20, after 3 warmup runs):
```
hyperfine -i -w3 -m20 "./bin/godot.linuxbsd.editor.x86_64 -e --path ~/New\ Game\ Project4.0/ --quit"
Benchmark 1: ./bin/godot.linuxbsd.editor.x86_64 -e --path ~/New\ Game\ Project4.0/ --quit
  Time (mean ± σ):      6.282 s ±  0.238 s    [User: 4.806 s, System: 0.485 s]
  Range (min … max):    5.332 s …  6.459 s    20 runs
```

So it does seem to shave off around 350 ms in average for me, though it's not a very reliable test. Interesting that the cold run also seems to be tangentially faster / same duration, so saving the cache doesn't seem to be slow.

---

The above test was running on AC. When running on battery, I get this:

### Before this PR

```
$ hyperfine -i -w3 -m20 "./bin/godot.linuxbsd.editor.x86_64.old -e --path ~/New\ Game\ Project4.0/ --quit"
Benchmark 1: ./bin/godot.linuxbsd.editor.x86_64.old -e --path ~/New\ Game\ Project4.0/ --quit
  Time (mean ± σ):      7.682 s ±  0.339 s    [User: 6.179 s, System: 0.630 s]
  Range (min … max):    6.612 s …  8.085 s    20 runs
```

### After this PR

```
$ hyperfine -i -w3 -m20 "./bin/godot.linuxbsd.editor.x86_64 -e --path ~/New\ Game\ Project4.0/ --quit"
Benchmark 1: ./bin/godot.linuxbsd.editor.x86_64 -e --path ~/New\ Game\ Project4.0/ --quit
  Time (mean ± σ):      7.538 s ±  0.354 s    [User: 6.052 s, System: 0.712 s]
  Range (min … max):    6.254 s …  7.929 s    20 runs
```

So here too it seems to give a 150 ms speedup.

**Comment by RandomShaper** [comment]
Thank you for the thorough review.

PS. Rebased now the PR dependency has been merged.

**Review by akien-mga — APPROVED** [review]

**Comment by RedworkDE** [comment]
If an outdated cache exists, the `class_list` is missing a few entries. Notably `@GlobalScope` being missing causes #76250.

```diff
diff --git a/class_list b/class_list
index fe8c67b9c65..2ec0a0a7c4e 100644
--- a/diff-tmp
+++ b/diff-tmp
@@ -1,5 +1,3 @@
-@GDScript
-@GlobalScope
 AABB
 AcceptDialog
 AESContext
@@ -54,7 +52,6 @@ Area2D
 Area2D.SpaceOverride
 Area3D
 Area3D.SpaceOverride
-Array
 ArrayMesh
 ArrayOccluder3D
 AspectRatioContainer
@@ -292,7 +289,6 @@ CylinderShape3D
 DampedSpringJoint2D
 Decal
 Decal.DecalTexture
-Dictionary
 DirAccess
 DirectionalLight2D
 DirectionalLight3D
@@ -402,7 +398,6 @@ FileDialog
 FileDialog.AccessEnum
 FileDialog.FileModeEnum
 FileSystemDock
-float
 FlowContainer
 FlowContainer.AlignmentMode
 FogMaterial
@@ -668,7 +663,6 @@ Node2D
 Node3D
 Node3D.RotationEditModeEnum
 Node3DGizmo
-NodePath
 Noise
 NoiseTexture2D
 Object
@@ -697,19 +691,10 @@ ORMMaterial3D
 OS
 OS.RenderingDriver
 OS.SystemDir
-PackedByteArray
-PackedColorArray
 PackedDataContainer
 PackedDataContainerRef
-PackedFloat32Array
-PackedFloat64Array
-PackedInt32Array
-PackedInt64Array
 PackedScene
 PackedScene.GenEditState
-PackedStringArray
-PackedVector2Array
-PackedVector3Array
 PacketPeer
 PacketPeerDTLS
 PacketPeerDtls.Status
@@ -771,7 +756,6 @@ PhysicsServer2D.PinJointParam
 PhysicsServer2D.ProcessInfo
 PhysicsServer2D.ShapeType
 PhysicsServer2D.SpaceParameter
-PhysicsServer2DExtension
 PhysicsServer2DManager
 PhysicsServer3D
 PhysicsServer3D.AreaBodyStatus
@@ -793,7 +777,6 @@ PhysicsServer3D.ProcessInfo
 PhysicsServer3D.ShapeType
 PhysicsServer3D.SliderJointParam
 PhysicsServer3D.SpaceParameter
-PhysicsServer3DExtension
 PhysicsServer3DManager
 PhysicsServer3DRenderingServerHandler
 PhysicsShapeQueryParameters2D
@@ -996,7 +979,6 @@ RichTextEffect
 RichTextLabel
 RichTextLabel.ListType
 RichTextLabel.MenuItems
-RID
 RigidBody2D
 RigidBody2D.CcdMode
 RigidBody2D.CenterOfMassModeEnum
@@ -1044,7 +1026,6 @@ ShapeCast2D
 ShapeCast3D
 short
 Shortcut
-Signal
 Skeleton2D
 Skeleton3D
 SkeletonIK3D
@@ -1095,8 +1076,6 @@ StreamPeerTCP
 StreamPeerTcp.Status
 StreamPeerTLS
 StreamPeerTls.Status
-String
-StringName
 StyleBox
 StyleBoxEmpty
 StyleBoxFlat
@@ -1217,7 +1196,6 @@ Upnp.UpnpResult
 UPNPDevice
 UpnpDevice.IgdStatusEnum
 ushort
-Variant
 VBoxContainer
 Vector2
 Vector2i
```

---

Also random thought, but I don't think switch between different build configurations (ie mono and standard version) of the same commit or having uncommitted changes invalidates the cache.

File: editor/editor_help.cpp [hunk 1]
```diff
@@ -45,6 +47,33 @@
...
+protected:
+	static void _bind_methods() {
+		ClassDB::bind_method(D_METHOD("set_version_hash", "version_hash"), &DocCache::set_version_hash);
+		ClassDB::bind_method(D_METHOD("get_version_hash"), &DocCache::get_version_hash);
+
+		ClassDB::bind_method(D_METHOD("set_classes", "classes"), &DocCache::set_classes);
+		ClassDB::bind_method(D_METHOD("get_classes"), &DocCache::get_classes);
+
+		ADD_PROPERTY(PropertyInfo(Variant::STRING, "version_hash"), "set_version_hash", "get_version_hash");
+		ADD_PROPERTY(PropertyInfo(Variant::DICTIONARY, "classes"), "set_classes", "get_classes");
```
Comment (raulsntos) [inline editor/editor_help.cpp:66]: PropertyInfo of property `classes` here uses the type **Dictionary** but the argument of setter `DocCache::set_classes` and the return type of getter `DocCache::get_classes` is **Array**. So there is a mismatch.

Comment (RandomShaper) [reply editor/editor_help.cpp:66]: Good catch, my friend. Fixed in #77782.
